packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -1210,6 +1451,58 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
       }
     }
 
+    if (!disableExtensions && this._spacesExtension) {
+      try {
+        namespaces = await this._spacesExtension.getSearchableNamespaces(options.namespaces);
+      } catch (err) {
+        if (Boom.isBoom(err) && err.output.payload.statusCode === 403) {
+          // The user is not authorized to access any space, return an empty response.
+          return SavedObjectsUtils.createEmptyFindResponse<T, A>(options);
+        }
+        throw err;
+      }
+      if (namespaces.length === 0) {
+        // The user is authorized to access *at least one space*, but not any of the spaces they requested; return an empty response.
+        return SavedObjectsUtils.createEmptyFindResponse<T, A>(options);
+      }
+    }
+
+    // We have to first do a "pre-authorization" check so that we can construct the search DSL accordingly
+    const spacesToPreauthorize = new Set(namespaces);
+    let typeToNamespacesMap: Map<string, string[]> | undefined;
+    let preAuthorizationResult: CheckAuthorizationResult<'find'> | undefined;
+    if (!disableExtensions && this._securityExtension) {
+      preAuthorizationResult = await this._securityExtension.checkAuthorization({
+        types: new Set(types),
+        spaces: spacesToPreauthorize,
+        actions: ['find'],
+      });
+      if (preAuthorizationResult.status === 'unauthorized') {
+        // If the user is unauthorized to find *anything* they requested, return an empty response
+        this._securityExtension.addAuditEvent({
+          action: AuditAction.FIND,
+          error: new Error(`User is unauthorized for any requested types/spaces.`),
+          // TODO: include object type(s) that were requested?
+          // requestedTypes: types,
+          // requestedSpaces: namespaces,
+        });
```

Comment:
So, just to understand, why are some audit logging delegated to the `checkAuthorization` call with the audit log hook, and some still directly managed by the SOR via `addAuditEvent`?

Reply:
Do you mean in enforceAuthorization function where we pass in an audit callback? I think the main reasons for this are...
- Keeps ownership of throwing the unauthorized error within the enforceAuthorization function.
- Allows flexibility for how audit events are handled (or not handled) without complicating the enforceAuthorization method and what it needs to know (decouples the extension from the specifics of each operation).

Bulk operations audit callbacks create an audit log for each object, so this can be implemented via the callback. Also there isn't always a clear/simple correlation between operation, action string, and audit action. Currently, the extension doesn't need to know anything about determining how to call addAuditEvent based on the operation, or which action string matches which audit action.

The alternative would be to either return all of the needed information from enforce and let the caller execute its own audit and throw handling, or to directly call the addAuditEvent from within the extension (which would require a mapping for operations to action string to audit action enumeration, or more input parameters).

Sorry for writing a book.

Reply:
> Do you mean in enforceAuthorization function where we pass in an audit callback

Yea sorry, that's what I meant

Thanks for the clarifications.